### ISSUE TITLE
cmd_c is sampled identically for TROT and CRAWL, but TROT can only reach ~half the range — and the front/rear achievable gap (~2.5-3x) is not in the target either

### ISSUE THREAD

**Issue by Ace2932** [Ace2932/Fennec#311]
Two related findings from the reward-chain audit, both about a **uniform target over a non-uniform plant**.

## 1. `cmd_c` ignores the gait

`env.py:519-527` (reset) and `:1087-1092` (step): `is_crawl` gates gait offsets, duty, the cmd_f band and the cmd-vx band — but `cmd_c` is always drawn `U(FOOTSWING_MIN 0.015, footswing_max 0.06)` with **no `is_crawl` branch**.

Measured with `probe_ik_swingref.py` at cmd_f = 1.4 Hz (inside `TROT_F_MIN/MAX`): the 4-leg-mean achieved peak **plateaus at ~2.5-2.7 cm for any cmd_c ≥ 0.04**, and even the best-case least-loaded single foot tops out at ~4.4-4.6 cm. So roughly **the top third of the sampled range is unreachable in trot timing**.

`env.py:256-261` justifies `FOOTSWING_MAX = 0.06` from a servo-envelope figure of ~6-7 cm — but that number traces to the **crawl-only** ceiling (`env.py:177`, the crawl-ceiling probe at duty 0.75, 0.3-0.5 Hz), not the much lower trot ceiling. The sampling code never distinguishes the two regimes.

Softer than a hard zero-gradient bug — the two-sided squared cost still has *some* slope — but across the top of the range it is close to a constant, non-actionable bill on trot envs. It also compounds #309: a target the plant cannot reach is exactly the shape that made the v5 clearance term impotent.

## 2. The swing target is identical for all four legs; the legs are not

`env.py:916-918`: `z_ref` / `cmd_c` / `W_SWINGREF` are the same for every leg.

Per-leg data from `probe_ik_swingref.py` at cmd_c = 0.05-0.06:

```
FL / FR   reach 2.9 - 4.6 cm
RL / RR   reach 1.2 - 1.9 cm      <- 2.5-3x less
```

And FL/RR are the **same diagonal pair** (they swing together, `GAIT_OFFSETS` env.py:173), so the asymmetry tracks **front vs rear position**, not swing/stance timing.

This may already have shown up from a different angle: `env.py:722-731`'s `w_carry` comment records *"probe after the contact/clearance fix: RR still carried 0.970, PINNED at ckpt12's value — zero gradient, not slow."* A rear leg structurally unable to reach a target its front counterpart can approach is a plausible root cause for a pinned carried-leg the codebase already flagged and could not explain.

Same class as #180 (front/rear treated identically where the geometry differs) — but in the **reward term** rather than an IK range.

## Suggested

* Branch `cmd_c`'s range on `is_crawl`, with the trot ceiling derived from the trot probe rather than inherited from the crawl figure.
* Consider a per-leg (or front/rear-keyed) swing target, or at minimum verify whether the rear legs' reachable range makes the uniform target unsatisfiable for them — and if so, whether that explains the pinned RR carry.

Both are measured open-loop; a trained policy may distribute effort differently. Worth confirming closed-loop before re-weighting anything.

**Comment by Ace2932** [comment]
## Confirmed in the source, and it is sharper than the original report

`env.py:462-465` — `sample_command()` gates the forward-velocity band on `is_crawl`:

```python
if is_crawl is not None:
    _c = is_crawl > 0.5
    lo = lo.at[0].set(jp.where(_c, CRAWL_VX_MIN, self._cmd_lo[0]))
    hi = hi.at[0].set(jp.where(_c, CRAWL_VX_MAX, self._cmd_hi[0]))
```

`env.py:525-527` — `cmd_c`, sampled in the same `info` dict literal ten lines later, is **not** gated on anything:

```python
"cmd_c": (jax.random.uniform(jax.random.fold_in(kc, 1), (),
                             minval=FOOTSWING_MIN, maxval=self._footswing_max)
          if self._heightmap else jp.asarray(BLIND_FOOTSWING)),
```

`cmd_f` is gated too (`:538`). So of the three per-episode commands v8 samples, two were made gait-aware and the third was not.

## The reasoning for the fix is already written in the file, one screen above

`env.py:192-198`, the comment justifying the vx gate:

> If a crawl env were commanded the trot's forward vx (up to +0.35 m/s) the progress + track rewards would demand a speed the slow schedule physically can't reach -> the tracking gradient would fight the gait, pulling the policy OFF the crawl schedule (the Task-1 carry-forward).

And the measured ceilings, `env.py:176-178`:

> the crawl-ceiling probe (`probe_crawl_ceiling.py` e6888d5) hit **6.7 cm STABLE, ~2x the trot's ~3 cm**, because a trot's 2-leg diagonal support sags under load.

`FOOTSWING_MAX = 0.06` (6 cm) is the upper bound for **both** gaits. So a trot env can be commanded a 6 cm footswing against a ~3 cm ceiling — up to 2x unreachable — which is the exact failure mode that paragraph describes, in the same reward pair (`w_swingref` tracks `z_ref = cmd_c*sin(pi*swing_frac)`), introduced by the same commit (#141), and left unfixed for this one command.

This is `fixed-the-instance-not-the-convention`: the justification was written once, applied to vx and `cmd_f`, and never carried to `cmd_c` even though the same sentence covers it. Note the direction — the crawl band is fine (6.7 cm ceiling > 0.06 max); it is **trot** that is over-commanded.

## Bearing on #309

#309 found the v7 swing-reference gate failing at the shipped `W_SWINGREF=100`, argmax at a=0.2 against a >=4.5 cm bar. If a large share of trot episodes carry a `cmd_c` the gait cannot reach, `w_swingref` is partly being asked to track an impossible reference, and `w_track` dominating it is what you would expect. These may be one defect, not two — worth resolving #311 before drawing conclusions from #309's weight sweep.

## Not fixing it blind

The contained fix mirrors the shipped vx gate: a `jp.where(_c, ...)` on `cmd_c`'s `maxval` with a trot ceiling near 0.03. But "~3 cm" is prose in a comment, not a probe artifact the way `CRAWL_VX_MAX` traces to `probe_crawl_ceiling.py` — and `probe_lift_ceiling.py` exists and can produce the real per-gait number. Getting the trot bound from the probe first is the difference between this fix and a second guessed constant.

**Comment by Ace2932** [comment]
## Measured the trot ceiling — and it argues against the obvious fix

Ran `probe_lift_ceiling.py` on current `main` (CPU, 57 s). It is already a **trot** probe — diagonal FL+RR / FR+RL phasing — so it answers this directly.

```
== TROT sweep — mean per-cycle peak foot_h ==
  T\a       0.2      0.4      0.6      0.8      1.0
  0.2s   1.27cm   1.27cm   1.34cm   1.37cm   1.40cm
  0.3s   1.36cm   1.65cm   2.30cm   2.46cm   2.41cm
  0.4s   1.35cm   2.08cm   3.23cm   3.29cm   3.34cm

BEST SUSTAINED (mean-per-cycle) TROT: 3.34 cm at a=1.0 T=0.4s
BEST SINGLE-CYCLE PEAK:               8.23 cm  (leg=FR, a=1.0 T=0.4s)
```

**The "~3 cm" prose in `env.py:176-178` checks out** — sustained trot tops out at 3.34 cm against `FOOTSWING_MAX = 0.06`. So a trot env can be commanded ~1.8x its own sustained ceiling, which is the core of this issue and it stands.

**But the single-cycle number breaks the clean version of the fix.** 8.23 cm on one cycle means the actuation chain *can* exceed 6 cm in a trot — it just cannot hold it with an open-loop sine. In this probe's own framing ("H-control vs H-valley", docstring) that is a **valley** result, not a ceiling: the servos are not the limit.

So clamping `cmd_c` to ~0.033 for trot would be wrong in kind. It would encode "an open-loop scripted sine cannot sustain this" as "the robot cannot do this", and a trained policy is not an open-loop sine — which is exactly the confound already noted against #309's probe. The gate would then permanently forbid a clearance the hardware demonstrably reaches.

**Revised read of this issue.** The defect is real but narrower than "trot is over-commanded": `cmd_c` is sampled with no gait coupling *at all*, while vx and `cmd_f` both are. The right fix is not a guessed trot cap — it is either
(a) couple `cmd_c`'s upper bound to something the policy can actually influence (gait period/amplitude), since the ceiling here is strongly T-dependent — 1.4 cm at T=0.2 s vs 3.3 cm at T=0.4 s, a 2.4x swing from period alone; or
(b) leave the range and accept that high-`cmd_c` trot episodes are a hard-exploration regime rather than an impossible one.

Either way it needs a closed-loop measurement, not this open-loop one. Not shipping a constant on this evidence.

## Side result: front/rear asymmetry has support

```
-- top 5 individual per-cycle peaks --
    8.23 cm  leg=FR
    5.98 cm  leg=FL
    5.84 cm  leg=FL
    5.61 cm  leg=FR
    5.61 cm  leg=FL
```

All five are **front** legs; no rear leg appears. Consistent with the front/rear gap this issue raises, and with the pinned RR carry. Caveat: the probe prints no per-leg means, so this is n=5 tail behaviour — it supports the direction, it does **not** confirm the ~2.5-3x figure. Getting that properly needs the probe extended to emit per-leg statistics.

**Comment by Ace2932** [comment]
## Correction: the front/rear gap is 1.88x, not the ~2.5-3x I claimed

The top-5 tail in my previous comment was n=5 and I should not have quoted a ratio off it. Extended the probe to aggregate the per-cycle records it already collects (PR #320 — reporting only, nothing re-runs). Pooled across all 15 amplitude x period combos:

```
FL: mean 2.42 cm   max 5.98 cm   n=270
FR: mean 2.40 cm   max 8.23 cm   n=260
RL: mean 1.26 cm   max 2.19 cm   n=260
RR: mean 1.31 cm   max 2.98 cm   n=270
FRONT mean 2.41 cm vs REAR mean 1.29 cm  ->  1.88x
```

**The asymmetry is real and left/right-symmetric** (FL~FR, RL~RR to within 0.05 cm across ~260 cycles each), so it is a front/rear structural effect, not noise or a per-leg sign bug.

## Two things this sharpens

**1. The rear ceiling is harder than the front's.** No rear leg exceeds **3 cm** on its single best cycle (RL 2.19, RR 2.98), while a front leg reaches 8.23. So the "8.23 cm proves the actuation chain can do it" caveat from my last comment applies to the **front** legs only — for the rear there is no such counter-example anywhere in 520 cycles.

**2. The problem is at the FLOOR, not just the ceiling.** `FOOTSWING_MIN = 0.015` (1.5 cm). Rear sustained mean is **1.29 cm — below the minimum commanded clearance.** Every value in the `cmd_c` range, including its smallest, is above what a rear leg sustains open-loop. That is a stronger statement than "the top of the range is unreachable", and it is a plausible mechanism for the pinned RR carry this issue cites (`env.py:725-726`: "RR still carried 0.970, PINNED... zero gradient, not slow") — a leg billed against a target it never reaches at any commanded value gets a gradient that never changes sign.

Still not shipping a constant on open-loop evidence — see the caveat in my previous comment, which applies unchanged. But the target of a fix looks different now: gating the *upper* bound per gait was the obvious move and would not touch this. The rear legs' problem is the whole range.

**Comment by Ace2932** [comment]
## The rear-leg clearance deficit has a named cause, and it is not the servos

Following up my per-leg measurement (front 2.41 cm vs rear 1.29 cm mean, 1.88x). That asymmetry is a **direct consequence of the end-keyed hfe range**, and the mechanism is clean.

`build_mjcf.py:48` — the two caps are *not* symmetric and are limited by different things:

```
HFE_FOLD = 0.873 rad = 50.0 deg   # conservative CHASSIS cap, TOWARD the trunk
HFE_EXT  = 1.501 rad = 86.0 deg   # mechanical: leg self-collision 93 deg (LA-19) - 7 deg margin
```

`hfe_range()` keys them by END, correctly (#144/#179):

```
FRONT (sx +1): (-86.0, +50.0) deg
REAR  (sx -1): (-50.0, +86.0) deg
```

**Lift needs NEGATIVE hfe** — `probe_lift_ceiling.py`'s own sign calibration picks `hfe_sign=-1, kfe_sign=-1` as max lift. And `+hfe` is the same world rotation on all four legs (MJCF axis `0 1 0`), so all four feet swing rearward on `+hfe`. Therefore `-hfe` swings every foot **forward** — which is *away* from the trunk at the front and *toward* it at the rear.

So the travel available **in the lift direction** is:

| end | limited by | travel |
|---|---|---|
| front | `HFE_EXT` (mechanical self-collision) | **86.0 deg** |
| rear | `HFE_FOLD` (conservative chassis cap) | **50.0 deg** |

**front/rear = 1.72x, against a measured clearance ratio of 1.88x.**

Physically: a rear leg lifts by swinging its foot *forward, under the body* — and that is precisely the direction the belly-clearance cap restricts. The front legs lift by swinging forward *away* from the trunk, where nothing is in the way until the leg hits itself at 93 deg.

I would not claim 1.72 *proves* 1.88 — clearance is not linear in hfe angle, kfe contributes, and the probe is open-loop. But a 1.72x travel ratio sitting under a 1.88x clearance ratio is not a coincidence, and it names the cause.

## Why this matters more than a per-leg curiosity

Three things turn out to be the same thing:

1. **This issue's rear-leg deficit.**
2. **Rear clearance sits below `FOOTSWING_MIN`** — 1.29 cm sustained against a 1.5 cm floor, so *every* commanded value is unreachable at the rear, not merely the top of the range.
3. **The stair-climb blocker.** `HFE_FOLD` is the same conservative fold cap already identified as the real stair blocker (chassis, **software-fixable**, with 13.79 cm of posture budget available) — not a servo limit, which the 3.13x lift-force margin already disproved.

## The headroom is documented, and I am not spending it

`hfe_range()`'s own comment records that against the posture-aware chassis bounds at haa 0 / kfe -109, the gate allows **67.9 deg** toward-trunk where the model stops at 50.0. If rear lift travel went 50 -> 67.9, the front/rear travel ratio falls from 1.72x to **1.27x**, and rear sustained clearance would plausibly clear `FOOTSWING_MIN`.

That is a change to a **safety cap that trained policies were learned under**, so it is not mine to make on the strength of an open-loop probe. Recording it as the evidence-backed candidate, with the number it would move and where the number came from.
